lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,253 @@
+	// a forward and doesn't provide a preimage.
+	ErrMissingPreimage = errors.New("missing preimage")
+
+	// ErrInterceptorAlreadyExists is an error returned when the a new stream
+	// is opened and there is already one active interceptor.
+	// The user must disconnect prior to open another stream.
+	ErrInterceptorAlreadyExists = errors.New("interceptor already exists")
+)
+
+// HtlcInterceptor is a bidirectional stream for streaming interception
+// requests to the caller.
+// Upon connection it does the following:
+// 1. Check if there is already a live stream, if yes it rejects the request.
+// 2. Regsitered a ForwardInterceptor
+// 3. Delivers to the caller every √√ and detect his answer.
+// It uses a local implementation of holdForwardsStore to keep all the hold
+// forwards and find them when manual resolution is later needed.
+func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
+	// We ensure there is only one interceptor at a time.
+	rpcInterceptor := newRPCInterceptor(stream)
+	if err := s.setRPCInterceptor(rpcInterceptor); err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	quit := make(chan struct{})
+	defer func() {
+		close(quit)
+		s.onInterceptorDisconnected()
+		wg.Wait()
+	}()
+
+	// this go routine reads from the client stream and and then signal to the
+	// main loop to handle the packet.
+	clientResolutionsChan := make(chan *ForwardHtlcInterceptResponse)
+	errChan := make(chan error)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			// Now that we have the response from the RPC client, send it to
+			// the responses chan.
+			select {
+			case clientResolutionsChan <- resp:
+			case <-quit:
+				return
+			case <-s.quit:
+				return
+			}
+		}
+	}()
+
+	// Our interceptor makes sure we hold the packet and then signal to the
+	// main loop to handle the packet. We only return true if we were able
+	// to deliver the packet to the main loop.
+	interceptedPacketsChan := make(chan htlcswitch.InterceptedForward)
+	fromSwitchInterceptor := func(packet htlcswitch.InterceptedForward) bool {
+		select {
+		case interceptedPacketsChan <- packet:
+			return true
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+	}
+
+	// Register our interceptor and make sure we resolve the held packets
+	// on disconnect.
+	interceptableForwarder := s.cfg.RouterBackend.InterceptableForwarder
+	interceptableForwarder.SetInterceptor(fromSwitchInterceptor)
+
+	for {
+		select {
+		case intercepted := <-interceptedPacketsChan:
+			rpcInterceptor.hold(intercepted)
+			log.Tracef("sending intercepted packet to client %v", intercepted)
+			// in case we couldn't forward we exit the loop and drain the
+			// current interceptor as this indicates on a connection problem.
+			if err := rpcInterceptor.forwardToClient(intercepted); err != nil {
```

Comment:
I think `hold` and `forwardToClient` can be merged into a single function? They are only called here in sequence

Reply:
Indeed, merged.